src/compiler/checker/lines_16000_18000.rs

```diff
@@ -132,6 +140,195 @@ impl TypeChecker {
+        TypeMapper::new_composite(mapper1, mapper2)
+    }
+
+    pub(super) fn make_merged_type_mapper(
+        &self,
+        mapper1: TypeMapper,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        TypeMapper::new_merged(mapper1, mapper2)
+    }
+
+    pub(super) fn combine_type_mappers(
+        &self,
+        mapper1: Option<TypeMapper>,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        if let Some(mapper1) = mapper1 {
+            self.make_composite_type_mapper(mapper1, mapper2)
+        } else {
+            mapper2
+        }
+    }
+
+    pub(super) fn merge_type_mappers(
+        &self,
+        mapper1: Option<TypeMapper>,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        if let Some(mapper1) = mapper1 {
+            self.make_merged_type_mapper(mapper1, mapper2)
+        } else {
+            mapper2
+        }
+    }
+
+    pub(super) fn instantiate_symbol(
+        &self,
+        mut symbol: Rc<Symbol>,
+        mapper: &TypeMapper,
+    ) -> Rc<Symbol> {
+        let links = self.get_symbol_links(&symbol);
+        let links = links.borrow();
+        if let Some(type_) = links.type_.as_ref() {
+            if !self.could_contain_type_variables(type_.clone()) {
+                return symbol;
+            }
+        }
+        let mut mapper = (*mapper).clone();
+        if get_check_flags(&symbol).intersects(CheckFlags::Instantiated) {
+            symbol = links.target.clone().unwrap();
+            mapper = self.combine_type_mappers(links.mapper.clone(), mapper);
+        }
+        let result = self.create_symbol(
+            symbol.flags(),
+            symbol.escaped_name().clone(),
+            Some(
+                CheckFlags::Instantiated
+                    | get_check_flags(&*symbol)
+                        & (CheckFlags::Readonly
+                            | CheckFlags::Late
+                            | CheckFlags::OptionalParameter
+                            | CheckFlags::RestParameter),
+            ),
+        );
+        if let Some(declarations) = &*symbol.maybe_declarations() {
+            result.set_declarations(declarations.clone());
+        }
+        let symbol_links = result.symbol_links();
+        let mut symbol_links_ref = symbol_links.borrow_mut();
+        symbol_links_ref.target = Some(symbol.clone());
+        symbol_links_ref.mapper = Some(mapper);
+        if let Some(value_declaration) = &*symbol.maybe_value_declaration() {
+            result.set_value_declaration(value_declaration.upgrade().unwrap());
+        }
+        Rc::new(result.into())
+    }
+
+    pub(super) fn instantiate_type(
+        &self,
+        type_: Option<Rc<Type>>,
+        mapper: Option<&TypeMapper>,
+    ) -> Option<Rc<Type>> {
+        if let Some(type_) = type_.as_ref() {
+            if let Some(mapper) = mapper {
+                return Some(self.instantiate_type_with_alias(type_.clone(), mapper, None, None));
+            }
+        }
+        type_
```

Comment:
Unless you're trying to mimic a larger if-statement, I'd write this with a match:

```rs
match (type_.as_ref(), mapper) {
    (Some(type_), Some(mapper)) => {
        return Some(self.instantiate_type_with_alias(type_.clone(), mapper, None, None))
    }
    _ => type_,
}
```

Reply:
Ok ya this isn't straying too far from the Typescript version (which uses a ternary `type && mapper ? ... : ...`) as a way of expressing conditional "if this/these are defined, do this, otherwise do this" I guess, updated to use `match`